Draft review of dotnet guide for code coverage import destined for Community guides (#1)
joe-tingsanchali-sonarsource/community-guides: dotnet-coverage-test-data-import-guide.md

* Visual Studio Code Coverage
* dotCover
* OpenCover/Coverlet

If you do not use any of the above code coverage tools, code coverage import will not work.

Understand the basics of...
* SonarScanner for MSBuild by reading its docs [here](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/), in case you are using it in your pipeline.
* Azure Devops - understand the integration between Azure DevOps and SonarCloud by reading [this guideline](https://azuredevopslabs.com//labs/vstsextend/sonarcloud/). =====> will add the reference to the yaml file for “Prepare SonarCloudAnalysis” step......

Note that SonarScanner for MSBuild has been renamed as SonarScanner for .NET.

# Troubleshooting steps

* Are you using a valid version of your code coverage tool?

* Are the reports in the right format? (need to check which file type matches which code coverage tool)

* Do the reports even exist?
 If you see the following, then ensure the report generation for your code coverage tool actually produced a file and has content in it.
 
     ```
     WARN: The Code Coverage report doesn’t contain any coverage data for the included files.
     ```
 

>     * Verify Indexed Paths Compared to Coverage File Paths

Reviewer: May be worth adding a note here concerning the /p:PathMap option of MSBuild, which is not supported for coverage files being parsed for SQ / SC

 (see https://community.sonarsource.com/t/code-coverage-is-not-shown-in-the-sonarcloud-io/15218/6)
* Do the reports contain valid information for the expected set of files? (give examples)

* Did you specify the correct report path property to import the report at the right step?

> When importing code coverage reports, you need to ensure that you specify the report path with your associated code coverage tool's property for the SonarScanner.MSBuild.exe `begin` command, as detailed in [Test Coverage](https://docs.sonarqube.org/latest/analysis/coverage/):

Reviewer: We also have a automatic detection and write of 
- sonar.cs.vscoveragexml.reportsPaths
- sonar.cs.vstest.reportsPaths

properties inside the Scanner for .NET, if proper trx / coverage files have been found in specific folders.

Author: @mickael-caro-sonarsource: Which folders are they? I can add that to make it more clear.

Reviewer: IIRC, we look at $(Agent.BuildDirectory)\TestResults first, then we fallback in $(Agent.TempDirectory). This is working only on a Azure DevOps environment.

Author: Good call! I see it in the [Code Coverage](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/#header-5) section. I already added in the introduction section, but I'll reference it again here as well:  look in $Build.SourcesDirectory path first then fallback to $Agent.TempDirectory.

> In a Azure DevOps / TFS environment, test files are automatically retrieved following this search
>
>  * Search for .trx files in any TestResults folder located under the $Build.SourcesDirectory path
>  * If not found, then a fallback search is made against $Agent.TempDirectory

    * sonar.cs.vscoveragexml.reportsPaths
        * Path to Visual Studio Code Coverage report. Multiple paths may be comma-delimited, or included via wildcards.
        * Ensure that you have converted the code coverage report from Binary into XML (see "Convert the Code Coverage Report from Binary into XML" section on our [[Coverage & Test Data] Generate Reports for C#, VB.net](https://community.sonarsource.com/t/coverage-test-data-generate-reports-for-c-vb-net/9871) Community guide)
    * sonar.cs.dotcover.reportsPaths
        * Path to dotCover coverage report.
    * sonar.cs.opencover.reportsPaths
        * Path to OpenCover coverage report.
    * sonar.vbnet.dotcover.reportsPaths
        * Path to dotCover coverage report.

>     * sonar.vbnet.opencover.reportsPaths

Reviewer 2: also `sonar.vbnet.vscoveragexml.reportsPaths`

        * Path to OpenCover coverage report.

* Did you specify the correct path to your code coverage file?

    Make sure you understand the regular expression for matching files and directories:
    |Symbol|Meaning|
    |---|---|
    |?|a single character|
    |*|any number of characters|
    |**|any number of directories|
    
    Here is a resource to understand some other examples: [Pattern matching guide](https://confluence.atlassian.com/fisheye/pattern-matching-guide-960155410.html)

    Examples of Correct Usage:
    

>     Right :o:

Reviewer 2: I find the red confusing for `right` - can this be a green circle?

Reviewer 2: or even better: ✅

    
    :o: `-d:"sonar.cs.vscoveragexml.reportsPaths=coveragexml"` (assuming you are running SonarScanner.MSBuild.exe in same directory as your *.coveragexml file)
    
    :o: `-d:"sonar.cs.dotcover.reportsPaths=**/*.html"`
    
    :o: `-d:"sonar.vbnet.opencover.reportsPaths=**/someNameCodeCoverage.xml"`
    
    Wrong :x:
    
    :x:  Don't point at directories but files
        `-d:"sonar.cs.vscoveragexml.reportsPaths=coverage/"`
        
    :x: Don't use absolute paths, favor paths relative to the project root directory
        `-d:"sonar.cs.vscoveragexml.reportsPaths=C:\users\joe\project\coverage\*.xml"`
        
    :x: Don't point files that are not under the project root directory, always generate coverage under this home directry
        `-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"`

* Do you know how to get debug logs?

    Add `/d:"sonar.verbose=true"` to the...
    * SonarScanner.MSBuild.exe `begin` command to get more detailed logs (TODO======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden)

>     * `SonarCloudPrepare` task in case you are using Azure DevOps

Reviewer 2: here you can provide the YAML and/or screenshot of the step

>         * For example: `SonarScanner.MSBuild.exe begin /k:"MyProject" /d:"sonar.verbose=true"`

Reviewer 2: this should be between `SonarScanner.MSBuild.exe `begin`` and ``SonarCloudPrepare` task`

>             The parsing and import of code coverage happens during the Scanner for MSBuild `end` command (the Azure Devops `SonarCloudAnalyze` task).

Reviewer 2: ```suggestion
The parsing and import of code coverage happens during the Scanner for MSBuild `end` command (the Azure Devops `SonarCloudAnalyze` task).
```

This is valid for both points, so should not be indented.

    Make sure you understand the regular expression for matching files and directories:
    |Symbol|Meaning|
    |---|---|
    |?|a single character|
    |*|any number of characters|
    |**|any number of directories|
    
    Here is a resource to understand some other examples: [Pattern matching guide](https://confluence.atlassian.com/fisheye/pattern-matching-guide-960155410.html)

    Examples of Correct Usage:
    
    **Right** ✅
    
    ✅ Do use glob patterns to match the ultimately created
    
    ```-d:"sonar.cs.vscoveragexml.reportsPaths=**/*.coveragexml”``` or
    
    ```-d:"sonar.cs.dotcover.reportsPaths=**/*.html"``` or
    
    ```-d:"sonar.vbnet.opencover.reportsPaths=**/someNameCodeCoverage.xml"```
    
    ✅ Values can be comma separated
    
    ```/d:"sonar.cs.vscoveragexml.reportsPaths=coverage/joe*.coveragexml,otherdir/olivier*.coveragexml"```
    
    ✅ `-` and `/` are valid flag markers
    
    ```/d:"sonar.verbose=true``` or
    ```-d:"sonar.verbose=true"``` 
  
    ✅ Report path is relative to the directory where you run SonarScanner for MSBuild 
    
    ```-d:"sonar.cs.vscoveragexml.reportsPaths=coveragexml"```
    
    **Wrong** ❌
    

>     ❌  Don't point at directories but files

Reviewer 2: ```suggestion
    ❌  Don't point at directories. **How to fix:** point to files.
```

    
    ```-d:"sonar.cs.vscoveragexml.reportsPaths=coverage/"```
        

>     ❌ Don't use absolute paths, favor paths relative to the project root directory

Reviewer 2: ```suggestion
    ❌ Don't use absolute paths. **How to fix:** favor paths relative to the project root directory.
```

    
    ```-d:"sonar.cs.vscoveragexml.reportsPaths=C:\users\joe\project\coverage\*.xml"```
        

>     ❌ Don't point files that are not under the project root directory, always generate coverage under this home directory

Reviewer 2: ```suggestion
    ❌ Don't point files that are not under the project root directory. **How to fix:** always generate the coverage report under the directory where you run the command from.
```

  
    ```-d:"sonar.cs.vscoveragexml.reportsPaths=..\coverage\*.xml"```

* Do you know how to get debug logs?

    * Add `/d:"sonar.verbose=true"` to the...

>         * SonarScanner.MSBuild.exe `begin` command to get more detailed logs (TODO======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden)

Reviewer 2: it's the `extraProperties`
```
- task: SonarCloudPrepare@1
  inputs:
    SonarCloud: 'sonarcloud'
    organization: 'joetingsanchali'
    scannerMode: 'MSBuild'
    projectKey: 'joetingsanchali_sonar-scanning-someconsoleapp'
    projectName: 'sonar-scanning-someconsoleapp'
    extraProperties: |
      sonar.cs.vstest.reportsPaths=$(Agent.TempDirectory)\**\*.trx
      sonar.cs.vscoveragexml.reportsPaths=$(Agent.TempDirectory)\**\*.coveragexml
      sonar.verbose=true
```

(I took it from [this awesome answer](https://community.sonarsource.com/t/need-help-for-unit-test-code-coverage-for-sonarcloud/32929/5?u=andrei_epure))

Author: That Joe guy sure knows something!

* sonar.vbnet.opencover.reportsPaths
    * Path to OpenCover coverage report.

Third, understand the basics of...

> * SonarScanner for MSBuild by reading its docs [here](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/), in case you are using it in your pipeline.

Reviewer 3: I'd anticipate the imminent change and use the work SonarScanner for .NET (formerly SonarScanner for MSBuild)

Author: Good point. I'll reference both then.

> * Azure Devops - understand the integration between Azure DevOps and SonarCloud by reading [this guideline](https://azuredevopslabs.com//labs/vstsextend/sonarcloud/). =====> will add the reference to the yaml file for “Prepare SonarCloudAnalysis” step......

Reviewer 3: Comment valid for several areas in the guide. Refer to the SonarCloud AND SonarQube equivalent reference (ie also mention SonarQubePrepareAnalysis on this case)

Author: Yea, this is my weak point: I haven't used SonarCloud enough to see the equivalence. I'll have to play with SonarCloud more.

Author: I've studied further the art of SonarCloud vs. SonarQube task differences in Azure DevOps and now I see the difference! Updated!

#### How to Get Debug Level Logs

> Add `/d:sonar.verbose=true` or `/d:"sonar.verbose=true"` to the...

Reviewer 3: Pick a single syntax here: I'd suggest to always use /d:"\<property\>=\<value\>" everywhere in the doc. That's the syntax that always work for any property

Author: True. Perhaps I can make a single sentence that mentions the equivalent syntax but stick with the double-quotes throughout?

EDIT: nevermind, I'll just stick to one syntax throughout.

* SonarScanner.MSBuild.exe `begin` command to get more detailed logs ======>should mention the alternative for Azure Devops integration, where the Scanner for MSBuild is hidden
* `SonarCloudPrepare` task in case you are using Azure DevOps

For example:

> `SonarScanner.MSBuild.exe begin /k:“MyProject” /d:sonar.verbose=true`

Reviewer 3: /d:"sonar.verbose=true"
